sql: fix performance regression due to increased hash allocations

arulajmani opened the pull request:
We recently increased allocations in the `appStats` structure when
adding support for transaction level statistics in #52704. This was
because the interactions with the `fnv` library were expensive in terms
of allocations. This patch aims to claw back the regression by:

- Using our own implementation of the FNV algorithm instead of the
library, which is significantly lighter weight (microbenchmarks below).
- Re-organizes the code to only construct the statement IDs (deemed the
expensive operation) if required.

When comparing the difference between the commit that introduced the
regression and the changes proposed by this diff, I got the following
improvements on the KV workload:

```
name             old ops/s   new ops/s   delta
kv95-throughput  34.5k ± 6%  35.7k ± 4%  +3.42%  (p=0.023 n=10+10)

```

Microbenchmarks for the new hashing algorithm (written/run by @knz):
```
name                     old time/op    new time/op    delta
ConstructStatementID-32     405ns ±17%      39ns ±12%  -90.34%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
ConstructStatementID-32      120B ± 0%       16B ± 0%  -86.67%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
ConstructStatementID-32      6.00 ± 0%      1.00 ± 0%  -83.33%  (p=0.008 n=5+5)
```

Closes #54515

Release note: None

arulajmani committed:
sql: fix performance regression due to increased hash allocations

We recently increased allocations in the `appStats` structure when
adding support for transaction level statistics in #52704. This was
because the interactions with the `fnv` library were expensive in terms
of allocations. This patch aims to claw back the regression by:

- Using our own implementation of the FNV algorithm instead of the
library, which is significantly lighter weight (microbenchmarks below).
- Reorganizing the code to only construct a statement ID from scratch
- if required.

When comparing the difference between the commit that introduced the
regression and the changes proposed by this diff, I got the following
improvements on the KV workload:

```
name             old ops/s   new ops/s   delta
kv95-throughput  34.5k ± 6%  35.7k ± 4%  +3.42%  (p=0.023 n=10+10)

```

Microbenchmarks for the new hashing algorithm (written/run by @knz):
```
name                     old time/op    new time/op    delta
ConstructStatementID-32     405ns ±17%      39ns ±12%  -90.34%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
ConstructStatementID-32      120B ± 0%       16B ± 0%  -86.67%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
ConstructStatementID-32      6.00 ± 0%      1.00 ± 0%  -83.33%  (p=0.008 n=5+5)
```

Closes #54515

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55214)
<!-- Reviewable:end -->

knz reviewed (commented):
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @solongordon)

___
*[pkg/roachpb/app_stats.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIns93gACVEv1XD85zS:-MIns93gACVEv1XD85zT:b-y0q382) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1401fbfddc1e6e577d3cb0b6292f8332641e25/pkg/roachpb/app_stats.go#L43)):*
> ```Go
> 	}
> 	b := util.FNV64ToByteArray(hash)
> 	return StmtID(*(*string)(unsafe.Pointer(&b)))
> ```

you don't know that  the byte array is not shared at this point. This cast is truly unsafe and should not be kept there.
___
*[pkg/sql/conn_executor_exec.go, line 1461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIns2DI0jPgiU2kAwdy:-MIns2DI0jPgiU2kAwdz:b9chmo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1401fbfddc1e6e577d3cb0b6292f8332641e25/pkg/sql/conn_executor_exec.go#L1461)):*
> ```Go
> 	commitLat := ex.phaseTimes.getCommitLatency()
> 
> 	key := txnKey(fmt.Sprintf("%x", util.FNV64ToByteArray(ex.extraTxnState.transactionStatementsHash)))
> ```

that is wayyyy too expensive. Just never use printf to compute keys or things on the hot path.
___
*[pkg/util/hash.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MInroG04G6k-qeL6nS5:-MInroG04G6k-qeL6nS6:b-i7149d) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1401fbfddc1e6e577d3cb0b6292f8332641e25/pkg/util/hash.go#L42)):*
> ```Go
> }
> 
> func FNV64ToByteArray(s0 uint64) []byte {
> ```

as per the previous discussion I'd encourage you to try make the stmt key an uint64 directly instead of a string

barring that, make this function return a string directly, otherwise you get weird allocation semantics with the unsafe cast.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Switched the statement keys/txn keys to be uint64s instead of strings. RFAL. 

CC-ing @dhartunian on this as this will probably affect the adminUI too. We recently discovered a significant performance regression because of increased string allocations. As a fix for this, I've updated the proto so that statement ID is now a uint64 instead of a string. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-:-MIu-Nzb3pbvj7Xj5o0Y:bydu6r3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @nvanbenschoten, and @solongordon)

___
*[pkg/roachpb/app_stats.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIns93gACVEv1XD85zS:-MItvEG28TV9YR34EMqU:b-dynbin) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1401fbfddc1e6e577d3cb0b6292f8332641e25/pkg/roachpb/app_stats.go#L43)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you don't know that  the byte array is not shared at this point. This cast is truly unsafe and should not be kept there.
</blockquote></details>

Not applicable anymore as I've made `StmtID` a uint64 instead of a string.
___
*[pkg/sql/conn_executor_exec.go, line 1461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIns2DI0jPgiU2kAwdy:-MItvN2B4EInnoTpjf9Y:bs8jjxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1401fbfddc1e6e577d3cb0b6292f8332641e25/pkg/sql/conn_executor_exec.go#L1461)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

that is wayyyy too expensive. Just never use printf to compute keys or things on the hot path.
</blockquote></details>

Same comment as above. 
___
*[pkg/util/hash.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MInroG04G6k-qeL6nS5:-MItvPwE1o1A5E4Ra4Ns:b-zcjgex) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d1401fbfddc1e6e577d3cb0b6292f8332641e25/pkg/util/hash.go#L42)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

as per the previous discussion I'd encourage you to try make the stmt key an uint64 directly instead of a string

barring that, make this function return a string directly, otherwise you get weird allocation semantics with the unsafe cast.
</blockquote></details>

Switched to uint64. 


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 13 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @solongordon)

___
*[pkg/sql/crdb_internal.go, line 956 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIxqc3gEm_ln9i_ZqUb:-MIxqc3gEm_ln9i_ZqUc:bz8dj1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/crdb_internal.go#L956)):*
> ```Go
> 				stmtIDsDatum := tree.NewDArray(types.String)
> 				for _, stmtID := range s.statementIDs {
> 					if err := stmtIDsDatum.Append(tree.NewDString(strconv.FormatUint(uint64(stmtID), 10))); err != nil {
> ```

Maybe add a TODO here to migrate the schema of the vtable to use an integer type.

(Not in this change though)


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 6 files at r1, 13 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @solongordon)

___
*[pkg/server/serverpb/status.proto, line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyHqJG52TM6EPfPsgm:-MIyHqJG52TM6EPfPsgn:b-xoixze) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/server/serverpb/status.proto#L893)):*
> ```protobuf
>   message CollectedStatementStatistics {
>     ExtendedStatementStatisticsKey key = 1 [(gogoproto.nullable) = false];
>    uint64 id = 3 [(gogoproto.customname) = "ID",
> ```

nit: is the spacing off here?
___
*[pkg/sql/app_stats.go, line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIMy7-h1RkZ2c1ufb:-MIyIMy7-h1RkZ2c1ufc:b4rrfi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L247)):*
> ```Go
> 	// is an expensive operation).
> 	s := a.getStatsForStmtWithKey(key, invalidStmtID, false /* createIfNonexistent */)
> 	if s == nil && createIfNonexistent {
> ```

What fraction of the time do we expect to hit this branch? Would it make sense to move `constructStatementIDFromStmtKey` under the lock and the `if !ok && createIfNonexistent {` branch to avoid locking twice?
___
*[pkg/sql/app_stats.go, line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIAEw-oQLXMbDvP-p:-MIyIAEw-oQLXMbDvP-q:b4sz5ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L257)):*
> ```Go
> 	key stmtKey, ID roachpb.StmtID, createIfNonexistent bool,
> ) *stmtStats {
> 	a.Lock()
> ```

Can these functions acquire a read lock if `!createIfNonexistent`?
___
*[pkg/sql/executor_statement_metrics.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIp_aE3iRYj3bGW2_:-MIyIp_aE3iRYj3bGW2a:bjmpo5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/executor_statement_metrics.go#L200)):*
> ```Go
> 	// by the internal executor, in which case the hash is uninitialized (=0), and
> 	// can therefore be safely ignored.
> 	if ex.extraTxnState.transactionStatementsHash != 0 {
> ```

Should we be comparing against `invalidStmtID` instead of `0` in a few spots?
___
*[pkg/util/hash.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyJ-W05xY0E1K5pohS:-MIyJ-W05xY0E1K5pohT:bbhwg1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/util/hash.go#L32)):*
> ```Go
> const fnvPrime = 1099511628211
> 
> // FNV64Init initializes a new hash function.
> ```

Very cool!

You might consider adding some type safety here to prevent misuse. Something like:
```go
type FNV64 struct {
    sum uint64
}

func NewFnv64() FNV64 {
    return FNV64{sum: fnvBase}
}

func (f *FNV64) Add(c uint64) {
    f.sum *= fnvPrime
	f.sum ^= c
}

func (f *FNV64) Sum() uint64 {
    return f.sum
}
```
Just an idea though, feel free to ignore.


<!-- Sent from Reviewable.io -->

dhartunian committed:
ui: update admin-ui-components to 0.1.17

This change upgrades the `admin-ui-components`
dependency. The latest version uses the modified
protobuf type for statements IDs in order to render
the Transactions Page Table.

No user-facing changes to any functionality were
made.

Release note: none

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @nvanbenschoten, and @solongordon)

___
*[pkg/server/serverpb/status.proto, line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyHqJG52TM6EPfPsgm:-MIzhotODVvOrFL0Xa9w:b-pc2i2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/server/serverpb/status.proto#L893)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: is the spacing off here?
</blockquote></details>

Seems so, fixed. 
___
*[pkg/sql/app_stats.go, line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIMy7-h1RkZ2c1ufb:-MIzhr5l3nbj-o83zA0Z:bs7wubq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L247)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What fraction of the time do we expect to hit this branch? Would it make sense to move `constructStatementIDFromStmtKey` under the lock and the `if !ok && createIfNonexistent {` branch to avoid locking twice?
</blockquote></details>

Done! 
___
*[pkg/sql/app_stats.go, line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIAEw-oQLXMbDvP-p:-MIzhtOH9XThoqFpr_se:b-9ad9aj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L257)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can these functions acquire a read lock if `!createIfNonexistent`?
</blockquote></details>

I think they should, though that would be a bigger change as if I switch `appStats` to have a RWLock, all other places that lock app stats should be updated as well. Would you be okay if I tackled that in a follow-up PR instead of this change?  
___
*[pkg/sql/crdb_internal.go, line 956 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIxqc3gEm_ln9i_ZqUb:-MIz0Ksy4Wwvboh5g9a6:b-5klu0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/crdb_internal.go#L956)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe add a TODO here to migrate the schema of the vtable to use an integer type.

(Not in this change though)
</blockquote></details>

Wouldn't that cause potential overflows as these IDs are unsigned integer? 
___
*[pkg/sql/executor_statement_metrics.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIp_aE3iRYj3bGW2_:-MIzlTEYCmbP5vGD1ypf:b-vq8ipu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/executor_statement_metrics.go#L200)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we be comparing against `invalidStmtID` instead of `0` in a few spots?
</blockquote></details>

Not applicable anymore because of the FNV64 struct switch below. 
___
*[pkg/util/hash.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyJ-W05xY0E1K5pohS:-MIz19J84QC0PlZUarCE:b2dcvxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/util/hash.go#L32)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Very cool!

You might consider adding some type safety here to prevent misuse. Something like:
```go
type FNV64 struct {
    sum uint64
}

func NewFnv64() FNV64 {
    return FNV64{sum: fnvBase}
}

func (f *FNV64) Add(c uint64) {
    f.sum *= fnvPrime
	f.sum ^= c
}

func (f *FNV64) Sum() uint64 {
    return f.sum
}
```
Just an idea though, feel free to ignore.
</blockquote></details>

Much cleaner, done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
I found an issue with our existing code when I was moving statement ID creation inside `getStatsForStmtWithKey`. `GetStatsForStmt` and everything calling above it need to know the statement ID regardless of wether a stats object entry is created or not. This is because:
- Statement stats collection could be turned off, but we still might be collecting transaction statistics. 
- A statement may not meet the latency threshold to be recorded, but the statement ID is still required to disambiguate transactions that comprise that statement. 

The error was that even though we were creating the statement ID in `getStatsForStmt`, we never really returned it. This meant that if `createIfNonExistent` was false (case 2, from above), we would be returning a nil stats struct, and run into a nil exception further up the stack. I've fixed this oversight from before as part of my change.  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-:-MJ-9K4C26ErVfsN8X4h:b-pt2pcl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @nvanbenschoten, and @solongordon)



<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with nits, but also would be interested in nathan's final words

> Reviewed 6 of 8 files at r3, 2 of 5 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-:-MJ1GoN49pX__-T9TVVP:bwo5j5g)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @solongordon)

___
*[pkg/sql/app_stats.go, line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIAEw-oQLXMbDvP-p:-MJ1GPJnAuYkLgCBxbc4:bt2vrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L257)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think they should, though that would be a bigger change as if I switch `appStats` to have a RWLock, all other places that lock app stats should be updated as well. Would you be okay if I tackled that in a follow-up PR instead of this change?  
</blockquote></details>

Sure, but maybe add a TODO here
(Also I'd recommend you file an issue if you're not already working on the followup PR - I'd expect that there will be other high-priority work for you to work on next, and it'd be nice not to lose track of this)
___
*[pkg/sql/crdb_internal.go, line 956 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIxqc3gEm_ln9i_ZqUb:-MJ1GeY77IwmbF7oBeKa:b-e9lskp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/crdb_internal.go#L956)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Wouldn't that cause potential overflows as these IDs are unsigned integer? 
</blockquote></details>

uint64 -> int64 is lossless, large positive values simply become negative
___
*[pkg/util/hash.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyJ-W05xY0E1K5pohS:-MJ1Ftj27e7vfdzUQFMP:b7en74g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/util/hash.go#L32)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Much cleaner, done.
</blockquote></details>


Pardon my ignorance Nathan, but doesn't the pointer receiver on the method cause the compiler to place the struct on the heap, and force an allocation that we were working hard to eliminate?
___
*[pkg/util/hash.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MJ1G7Ej1BKz8QhXtCIF:-MJ1G7Ej1BKz8QhXtCIG:b-kunz2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6e7925a9c21e746807b7ba733cd3317e9b8d81a/pkg/util/hash.go#L38)):*
> ```Go
> 
> // NewFNV64 initializes a new FNV64 hash state.
> func NewFNV64() FNV64 {
> ```

a constructor that returns a value would be called "Make", not "New"


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 8 files at r3, 3 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @solongordon)

___
*[pkg/server/serverpb/status.proto, line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyHqJG52TM6EPfPsgm:-MJ2B7zTD-P1MBLWHxf5:b-x3gbo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/server/serverpb/status.proto#L893)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Seems so, though 
</blockquote></details>

I don't think this was fixed. It appears that there are only 3 spaces before the uint64.
___
*[pkg/sql/app_stats.go, line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIMy7-h1RkZ2c1ufb:-MJ2Bmum74liocmaib-4:bqgc5mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L247)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Done! 
</blockquote></details>

I'm not sure that this was done. We're still calling `getStatsForStmtWithKey` twice. It's not clear to me why we need to do that. Are we afraid of creating the stmtId under the lock because it's expensive? If so then why are we ok with allocating the stmtStats struct?

Or maybe this has to do with the fact that you want to return a stmtID even if `if !ok && !createIfNonexistent {`. If so, there are a few places around here that could use comments to describe what's going on. For instance, `recordStatement` should mention that it can never return an invalid statement ID. So should `getStatsForStmt`, even though it can return a nil `stmtStats` object.
___
*[pkg/sql/executor_statement_metrics.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIp_aE3iRYj3bGW2_:-MJ2D_Zi5uqTipHsxz2t:bbp9lt3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/executor_statement_metrics.go#L200)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Not applicable anymore because of the FNV64 struct switch below. 
</blockquote></details>

Do we still need `invalidStmtID` in that case?
___
*[pkg/util/hash.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyJ-W05xY0E1K5pohS:-MJ2Dy__EzEOUx8H1XUz:bmpc92) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/util/hash.go#L32)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Pardon my ignorance Nathan, but doesn't the pointer receiver on the method cause the compiler to place the struct on the heap, and force an allocation that we were working hard to eliminate?
</blockquote></details>

The methods should all be inlined because they all have trivial complexity, so it shouldn't. And even if they aren't inlined, escape analysis should determine that they don't let the receiver escape. But it's worth checking. You can use the `goescape` script from https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072868/Rando+scripts to do so. Run that and see if a call to the constructor results in a `"moved to heap"` line.

If you don't see anything, try making the following changes (which should cause the struct to escape) to confirm that you would detect a heap allocation if one existed.
```go
type FNV64 struct {
    sum   uint64
    recur *FNV64
}

func (f *FNV64) Add(c uint64) {
    f.sum *= fnvPrime
    f.sum ^= c
    f.recur = f
}
```
___
*[pkg/util/hash.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MJ1G7Ej1BKz8QhXtCIF:-MJ2DkMdEzThxSKS9_fL:bc2s38x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6e7925a9c21e746807b7ba733cd3317e9b8d81a/pkg/util/hash.go#L38)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

a constructor that returns a value would be called "Make", not "New"
</blockquote></details>

Yep, @knz is right about this. My mistake, thanks for catching!
___
*[pkg/util/hash.go, line 55 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MJ2DvSa1FsLgTIG8CZH:-MJ2DvSa1FsLgTIG8CZI:b-izygz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a4151aa039cc6bc57a49b1fcc7563247cfcbd91/pkg/util/hash.go#L55)):*
> ```Go
> }
> 
> // Sum returns the hash value accumulated so now.
> ```

"to now"?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Addressed all outstanding comments, should be good for another look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-:-MJ2cQ_b1DNCqrwuI0I5:b-8t2hon)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @nvanbenschoten, and @solongordon)

___
*[pkg/server/serverpb/status.proto, line 893 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyHqJG52TM6EPfPsgm:-MJ2Lcmk8h0pWB5UClv5:brddx6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/server/serverpb/status.proto#L893)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't think this was fixed. It appears that there are only 3 spaces before the uint64.
</blockquote></details>

Sorry, I misunderstood which spacing you meant earlier. Done now. 
___
*[pkg/sql/app_stats.go, line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIMy7-h1RkZ2c1ufb:-MJ2LpnM4_nb1wkgvNwb:beri005) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L247)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not sure that this was done. We're still calling `getStatsForStmtWithKey` twice. It's not clear to me why we need to do that. Are we afraid of creating the stmtId under the lock because it's expensive? If so then why are we ok with allocating the stmtStats struct?

Or maybe this has to do with the fact that you want to return a stmtID even if `if !ok && !createIfNonexistent {`. If so, there are a few places around here that could use comments to describe what's going on. For instance, `recordStatement` should mention that it can never return an invalid statement ID. So should `getStatsForStmt`, even though it can return a nil `stmtStats` object.
</blockquote></details>

It's the latter, I revisited this after my initial comment which made the comment above outdated. I'll leave comments around here that paint the picture. 
___
*[pkg/sql/app_stats.go, line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIAEw-oQLXMbDvP-p:-MJ2TFeu8hsXGpImWRsp:bguhp80) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/app_stats.go#L257)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Sure, but maybe add a TODO here
(Also I'd recommend you file an issue if you're not already working on the followup PR - I'd expect that there will be other high-priority work for you to work on next, and it'd be nice not to lose track of this)
</blockquote></details>

Filed #55285, and added a TODO in the `appStats` struct.
___
*[pkg/sql/crdb_internal.go, line 956 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIxqc3gEm_ln9i_ZqUb:-MJ2OH8vBDyxejXgz30z:b-s6eh0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/crdb_internal.go#L956)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

uint64 -> int64 is lossless, large positive values simply become negative
</blockquote></details>

Yes, but I'm slightly hesitant as then we'd be displaying different values in the admin UI and crdb_internal.node_transaction_statistics. Also, negative numbers for an ID field seem slightly weird from a UX perspective. 

I filed https://github.com/cockroachdb/cockroach/issues/55284 to capture the discussion, and I can make the change in the followup if my UX concerns don't stand. Added a TODO referencing the issue as well.
___
*[pkg/sql/executor_statement_metrics.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyIp_aE3iRYj3bGW2_:-MJ2OyJI8XnRJ8w_ZR0e:b-o8sxme) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/sql/executor_statement_metrics.go#L200)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we still need `invalidStmtID` in that case?
</blockquote></details>

Yeah, one other place still uses it. 
___
*[pkg/util/hash.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MIyJ-W05xY0E1K5pohS:-MJ2abPK253KNbcGEBsZ:b-dms59n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bc4c27ce0dbb7cc154e3929128869e40f51a56/pkg/util/hash.go#L32)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The methods should all be inlined because they all have trivial complexity, so it shouldn't. And even if they aren't inlined, escape analysis should determine that they don't let the receiver escape. But it's worth checking. You can use the `goescape` script from https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072868/Rando+scripts to do so. Run that and see if a call to the constructor results in a `"moved to heap"` line.

If you don't see anything, try making the following changes (which should cause the struct to escape) to confirm that you would detect a heap allocation if one existed.
```go
type FNV64 struct {
    sum   uint64
    recur *FNV64
}

func (f *FNV64) Add(c uint64) {
    f.sum *= fnvPrime
    f.sum ^= c
    f.recur = f
}
```
</blockquote></details>

There's no `moved to heap` line so I think we should be good. 

The code snippet you posted also doesn't result in a `moved to heap` line, presumably because it's a self reference? The following does though: 

```
func (f *FNV64) Add(c uint64) {
	f.sum *= fnvPrime
	f.sum ^= c
	a := MakeFNV64()
	f.recur = &a
}
```
___
*[pkg/util/hash.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MJ1G7Ej1BKz8QhXtCIF:-MJ2OqQfDsyGeZdvT75h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6e7925a9c21e746807b7ba733cd3317e9b8d81a/pkg/util/hash.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yep, @knz is right about this. My mistake, thanks for catching!
</blockquote></details>

Done.
___
*[pkg/util/hash.go, line 55 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-MJ2DvSa1FsLgTIG8CZH:-MJ2TinO0toPIIlpnFkM:bfwhuz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a4151aa039cc6bc57a49b1fcc7563247cfcbd91/pkg/util/hash.go#L55)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"to now"?
</blockquote></details>

so far/till now = so now :sweat:. Thanks for catching, fixed


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks for working through this!

> Reviewed 10 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55214#-:-MJ2f87Q5X7_gJgXJPOR:byqeuyi)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @nvanbenschoten, and @solongordon)



<!-- Sent from Reviewable.io -->

arulajmani commented:
Thanks for the reviews and help with this issue @nvanbenschoten @knz 

bors r=nvanbenschoten, knz
